File: Google Code Jam/GCJ2021/Qualification Round 2021/Moons_and_Umbrellas.py
def calculate_amount(strng):
    amount = 0
    for i in range(len(strng)-1):
        if strng[i:i+2] == 'CJ':
            amount = amount + x
        elif strng[i:i+2] == 'JC':
            amount = amount + y
    return(amount)

for t in range(int(input())):
    x,y,s = map(str,input().split())
    x = int(x)
    y = int(y)
    amount = 0
    temp_amt = 0
    i = 0
    ans = 0
    count_C = s.count('C')
    count_J = s.count('J')
    count = s.count('?')

    ssc = ""
    ssj = ""
    anscj = 0
    ansjc = 0

    if count == 0:
        ans = calculate_amount(s)
    elif count_C == 0:
        s = s.replace('?','J')
        ans = calculate_amount(s)
    elif count_J == 0:
        s = s.replace('?','C')
        ans = calculate_amount(s)
    else:
        while (i != len(s) - 1):
            if (s[i] == 'C' and s[i+1] == 'J'):
                ans = ans + x
                i += 1
            elif (s[i] == 'J' and s[i+1] == 'C'):
                ans = ans + y
                i += 1
            elif (s[i] == 'C' or s[i] == 'J'):
                i += 1
            else:
                ssc = s
                ssj = s
                anscj = ans
                ansjc = ans
                for j in range(i,len(s)):
                    if (s[j] != '?'):
                        ssc = ssc.replace(ssc[j],s[j],1)
                        ssj = ssj.replace(ssj[j],s[j],1)
                        if (ssc[j-1] == 'J' and ssc[j] == 'C'):
                            ansjc += y
                        elif (ssj[j-1] == 'C' and ssj[j] == 'J'):
                            anscj += x
                        elif (ssc[j-1] == 'C' and ssc[j] == 'J'):
                            ansjc += x
                        elif (ssj[j-1] == 'J' and ssj[j] == 'C'):
                            anscj +=y
                            i = j
                            break
                    else:
                        ssc = ssc.replace('?','C',1)
                        ssj = ssj.replace('?','J',1)
                        if (ssc[j-1] == 'J' and ssc[j] == 'C'):
                            ansjc += y
                        elif (ssj[j-1] == 'C' and ssj[j] == 'J'):
                            anscj += x
                        elif (ssc[j-1] == 'C' and ssc[j] == 'J'):
                            ansjc += x
                        elif (ssj[j-1] == 'J' and ssj[j] == 'C'):
                            anscj += y
                if ansjc < anscj:
                    s = ssc
                    ans = ansjc
                else:
                    s = ssj
                    ans = anscj

    print('case #'+str(t+1)+':',ans)

File: Google Code Jam/GCJ2021/Qualification Round 2021/Moons_and_Umbrellas.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long int ll;
typedef unsigned long long int ull;
typedef long double ld; //%Lf

#define MOD 1000000007


int main(void) {
  int t, tt;
  cin >> t;
  for (tt = 1; tt <= t; tt++) {
    int i = 0, j = 0, n, x, y, ans = 0, anscj, ansjc;
    string s, prev, ssc = "", ssj = "";
    cin >> x >> y >> s;
    for (i = 0; i < s.size();) {
      if (s[i] == 'C' && s[i + 1] == 'J') {
        ans += x;
        i++;
      } else if (s[i] == 'J' && s[i + 1] == 'C') {
        ans += y;
        i++;
      } else if (s[i] == 'C' || s[i] == 'J') {
        // prev = s[i];
        i++;
      } else {
        // if (s[i + 1] == '?') {
          ssc = s;
          ssj = s;
          anscj = ans;
          ansjc = ans;
          for (j = i; j < s.size(); j++) {
            if (s[j] != '?') {
              ssc[j] = s[j];
              ssj[j] = s[j];
              if (ssc[j - 1] == 'J' && ssc[j] == 'C') {
                ansjc += y;
              }
              if (ssj[j - 1] == 'C' && ssj[j] == 'J') {
                anscj += x;
              }
              if (ssc[j - 1] == 'C' && ssc[j] == 'J') {
                ansjc += x;
              }
              if (ssj[j - 1] == 'J' && ssj[j] == 'C') {
                anscj += y;
              }
              i = j;
              break;
            } else {
              ssc[j] = 'C';
              ssj[j] = 'J';
              if (ssc[j - 1] == 'J' && ssc[j] == 'C') {
                ansjc += y;
              }
              if (ssj[j - 1] == 'C' && ssj[j] == 'J') {
                anscj += x;
              }
              if (ssc[j - 1] == 'C' && ssc[j] == 'J') {
                ansjc += x;
              }
              if (ssj[j - 1] == 'J' && ssj[j] == 'C') {
                anscj += y;
              }
            }
          }
          if (ansjc < anscj) {
            s = ssc;
            ans = ansjc;
          } else {
            s = ssj;
            ans = anscj;
          }
        // }
      }
    }

    // cout << "Case #" << tt << ": " << s << " " << ans << endl;
    cout << "Case #" << tt << ": " << ans << endl;
  }
  return 0;
}
